lib/status.h

```diff
@@ -0,0 +1,159 @@
+#ifndef LIB_STATUS_H_
+#define LIB_STATUS_H_
+
+#include <memory>
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include "lib/invariant.h"
+
+namespace error {
+
+// Numeric constants describing common classes of errors.
+//
+// Note that the following helpers exist to check for common conditions:
+// IsOk() and IsError().
+enum class Code {
```

Comment:
I'm ok with this as is. But: does it make sense to define all those values already? Not convinced it is a particularly good collection.

Does it make sense to make a Code actually be two values, a pointer to a namespace object, and an actual integer value? Good idea, bad idea?

Concern: if the array remains generic, usefulness will be limited. If the array grows over time, than it'll become a single giant array for random error codes in ebplane. Also not good.

I have seen implementations using a namespace object, which were good and very functional (not the one in the status you are familiar with).
Go 2.0 also has a way to wrap an error inside another error, which is *extremely* nice and handy, you can do something like Wrap(ABORTED, othererror).

Eg, you have a timeout that causes a function to fail, you can have a "dataplane programming error" that embeds a "timeout error". Which is very convenient for debugging.

Reply:
I'd actually prefer to use a opaque handle type (actually my next pull request) and allow arbitrary values to be stored in Code, with the single convention being that zero implies success. If you don't mind I'll switch to this approach.

As for error namespaces, I think I view them as a way to decentralize the definition of constants by allocating "ranges" of values in bulk (each namespace implies some combination of high order bits).

The values are just taken from a limited version of util::Status that google open sourced years ago. For me I consider a direct embedding of errno to be most useful, but other applications have different needs.